### DISCUSSION TITLE
No_std support

### DISCUSSION THREAD

**Discussion by ghost** [parasyte/pixels#283] [Q&A] [answered]
Does this crate support being used in a program with no_std (such as a kernel)?
I'm wondering because I am trying to search for a crate to help me draw text on a framebuffer provided by a UEFI firmware

**Comment by parasyte** [comment, marked as answer]
No, unfortunately this crate will not work for your use case. We need a GPU for scaling (at a minimum) and our current GPU layer provides a Vulkan-like interface across all supported backends.

I'm not sure what kind of interface you would expect in a kernel, since presumably you wouldn't need a GPU or high-level interface to it like Vulkan. I suspect the standard VGA interface is what you're looking for? Specifically, this is what I'm thinking of: https://wiki.osdev.org/Drawing_In_a_Linear_Framebuffer and https://wiki.osdev.org/VGA_Resources

If instead you do expect to have a GPU and driver, then `pixels` could potentially be an option for you. At that point, we would need to make `wgpu` no_std. But it has been mentioned that they cannot do this: https://github.com/gfx-rs/wgpu/discussions/2278#discussioncomment-1790365

Alternatives you might consider:

- [`vga`](https://crates.io/crates/vga) offers low-level access to VGA registers and structures. Also provides an API for pixel plotting.
- [`vga-figures`](https://crates.io/crates/vga-figures) built on `vga` and provides drawing operations for basic shapes.

**Reply by ghost** [reply]
> No, unfortunately this crate will not work for your use case. We need a GPU for scaling (at a minimum) and our current GPU layer provides a Vulkan-like interface across all supported backends.
> 
> 
> 
> I'm not sure what kind of interface you would expect in a kernel, since presumably you wouldn't need a GPU or high-level interface to it like Vulkan. I suspect the standard VGA interface is what you're looking for? Specifically, this is what I'm thinking of: https://wiki.osdev.org/Drawing_In_a_Linear_Framebuffer and https://wiki.osdev.org/VGA_Resources
> 
> 
> 
> If instead you do expect to have a GPU and driver, then `pixels` could potentially be an option for you. At that point, we would need to make `wgpu` no_std. But it has been mentioned that they cannot do this: https://github.com/gfx-rs/wgpu/discussions/2278#discussioncomment-1790365
> 
> 
> 
> Alternatives you might consider:
> 
> 
> 
> - [`vga`](https://crates.io/crates/vga) offers low-level access to VGA registers and structures. Also provides an API for pixel plotting.
> 
> - [`vga-figures`](https://crates.io/crates/vga-figures) built on `vga` and provides drawing operations for basic shapes.

It is because I am given a basic framebuffer and I wanted an api to draw text for me on that framebuffer
